cirq/ops/pauli_sum_exponential.py

```diff
@@ -0,0 +1,131 @@
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality(approximate=True)
+class PauliSumExponential:
+    """Represents operator defined by exponential of a PauliSum.
+
+    Given a hermitian/anti-hermitian PauliSum PS_1 + PS_2 + ... + PS_N, this
+    class returns an operation which is equivalent to
+    exp(j * exponent * (PS_1 + PS_2 + ... + PS_N)).
+
+    This class currently supports only exponential of commuting Pauli Terms.
+    """
+
+    def __init__(self,
+                 pauli_sum_like: 'cirq.PauliSumLike',
+                 exponent: Union[int, float, sympy.Basic] = 1,
+                 *,
+                 is_anti_hermitian=False):
+        pauli_sum = linear_combinations.PauliSum.wrap(pauli_sum_like)
+        if not PauliSumExponential._all_pauli_strings_commute_(pauli_sum):
+            raise ValueError(
+                "PauliSumExponential defined only for commuting pauli sums.")
+        self._exponent = exponent
+        self._is_anti_hermitian = is_anti_hermitian
+        if is_anti_hermitian:
+            self._multiplier = -1j
+        else:
+            self._multiplier = 1
+        for pauli_string in pauli_sum:
+            normalised_coeff = complex(pauli_string.coefficient *
+                                       self._multiplier)
+            if normalised_coeff.imag != 0:
+                raise ValueError(
+                    pauli_string,
+                    "PauliSum should be either hermitian or anti-hermitian.")
+        self._pauli_sum = pauli_sum
+
+    @property
+    def qubits(self) -> Tuple['cirq.Qid', ...]:
+        return self._pauli_sum.qubits
+
+    def _value_equality_values_(self) -> Any:
+        return (self._pauli_sum, self._exponent)
+
+    def with_qubits(self, *new_qubits: 'cirq.Qid') -> 'PauliSumExponential':
+        return PauliSumExponential(self._pauli_sum.with_qubits(*new_qubits),
+                                   self._exponent,
+                                   is_anti_hermitian=self._is_anti_hermitian)
+
+    def _resolve_parameters_(self,
+                             param_resolver: 'cirq.ParamResolverOrSimilarType'
+                            ) -> 'PauliSumExponential':
+        return PauliSumExponential(self._pauli_sum,
+                                   exponent=protocols.resolve_parameters(
+                                       self._exponent, param_resolver),
+                                   is_anti_hermitian=self._is_anti_hermitian)
+
+    @staticmethod
+    def _all_pauli_strings_commute_(pauli_sum: 'cirq.PauliSum') -> bool:
+        for x in pauli_sum:
+            for y in pauli_sum:
+                if not protocols.commutes(x, y):
+                    return False
+        return True
+
+    def __iter__(self) -> Iterator['cirq.PauliStringPhasor']:
+        for pauli_string in self._pauli_sum:
+            theta = pauli_string.coefficient * self._multiplier
+            theta *= self._exponent / np.pi
+            if isinstance(theta, complex):
```

Comment:
Why are we just simply truncating here? Also, how can this be complex? Didn't we ensure in the constructor that for all coeffs coeff * multiplier is real? Also self._exponent should be real, right?

Reply:
Let res = coeff * multipler. We ensured that res is real (i.e. res.imag ~= 0) but the type of the object res is still complex and hence we simply truncate it to get the real part because the exponent passed to PauliStringPhasor should be float.

Reply:
Sounds good.